org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/compiler/XbaseCompiler.java

```diff
@@ -644,11 +644,40 @@ protected JvmTypeReference resolveSynonymType(JvmSynonymTypeReference reference,
 	/**
 	 * @since 2.18
 	 */
-	protected void appendFinallyWithResources(List<XVariableDeclaration> resources, ITreeAppendable b) {
-		for (int i = resources.size() - 1; i >= 0; i--) {
-			XVariableDeclaration res = resources.get(i);
-			String resName = getVarName(res, b);
-			b.newLine().append("if (" + resName + " != null) ").append(resName + ".close();");
+	protected void appendFinallyWithResources(XTryCatchFinallyExpression expr, List<XVariableDeclaration> resources,
+			ITreeAppendable b) {
+		if (expr != null && !resources.isEmpty()) {
+			// Invoking the close method might throw. Hence, we collect those
+			// Throwables and propagate them later on.
+			String throwablesStore = b.declareSyntheticVariable(Tuples.pair(expr, "_catchedThrowables"), "_ts");
+			b.newLine().append("java.util.List<Throwable> ");
```

Comment:
Instead of using FQN in generated code, we should use types instead so that the import manager can take care of them

Reply:
Done.

Reply:
Done.